### PR TITLE
feat: phase4

### PR DESCRIPTION
**Pull request by nugaon**
Introducing Phase4 storage incentives.

The PR is a bit bloated so it may be better to do the remaining test/refactor related tasks in other PRs, such as:
- CAC inclusion proof related tests are done by the fixtures generated by Bee clients. The missing part is to write SOC related and other tests.
- Do styling.
- Refactoring code (rid of state var usage where it is not intended, etc.).

### CODE REVIEW

File: src/Redistribution.sol [hunk 1]
```diff
@@ -101,13 +148,23 @@ contract Redistribution is AccessControl, Pausable {
     // Maximum value of the keccack256 hash.
     bytes32 MaxH = bytes32(0x00000000000000000000000000000000ffffffffffffffffffffffffffffffff);
 
+
+    // alpha=1% k=16 n=1M
+    uint256 public constant SAMPLE_MAX_VALUE = 94730660000000000000000000000000000000000000000000000000000000000000000000;
```
Comment (significance) [inline src/Redistribution.sol:153]: SAMPLE_MAX_VALUE is determined to be 1.284401 · 1072 in the storage incentives paper

Comment (nugaon) [reply src/Redistribution.sol:153]: yeah, I see it has changed. thanks for the notice!

**Comment by n00b21337** [comment]
I added all of the above suggestions here in this PR so we can easily choose when and how we want to apply them
https://github.com/ethersphere/storage-incentives/pull/164

File: src/PostageStamp.sol [hunk 2]
```diff
@@ -99,8 +99,8 @@ contract PostageStamp is AccessControl, Pausable {
 
     // Lottery pot at last update.
     uint256 public pot;
-
-    // blocks in 24 hours ~ 24 * 60 * 60 / 5 = 17280
+
```
Comment (mrekucci) [inline src/PostageStamp.sol:102]: Unnecessary white space, the same is below.

File: test/Redistribution.test.ts [hunk 3]
```diff
@@ -35,8 +50,10 @@ const reveal_nonce_f = '0xf4153f4153f4153f4153f4153f4153f4153f4153f4153f4153f415
 
 let node_1: string;
 const overlay_1 = '0xa6f955c72d7053f96b91b5470491a0c732b0175af56dcfb7a604b82b16719406';
+const overlay_1_n_25 = '0x676766bbae530fd0483e4734e800569c95929b707b9c50f8717dc99f9f91e915';
 const stakeAmount_1 = '100000000000000000';
 const nonce_1 = '0xb5555b33b5555b33b5555b33b5555b33b5555b33b5555b33b5555b33b5555b33';
+const nonce_1_n_25 = '0x00000000000000000000000000000000000000000000000000000000000325dd';
 const hash_1 = '0xb5555b33b5555b33b5555b33b5555b33b5555b33b5555b33b5555b33b5555b33';
```
Comment (mrekucci) [inline test/Redistribution.test.ts:57]: This issue reported by linter should be fixed.

File: test/Redistribution.test.ts [hunk 4]
```diff
@@ -634,65 +701,82 @@ describe('Redistribution', function () {
...
         });
       });
 
       describe('two commits with equal stakes', async function () {
         let priceOracle: Contract;
         let r_node_1: Contract;
-        let r_node_2: Contract;
+        let r_node_5: Contract;
         let currentRound: number;
+        let copyBatchTx: any;
```
Comment (mrekucci) [inline test/Redistribution.test.ts:730]: This issue reported by the checker should be fixed.

File: test/util/tools.ts [hunk 5]
```diff
@@ -105,6 +141,81 @@ async function mineOverlaysInDepth(
...
+    await mineNBlocks(PHASE_LENGTH - roundBlocks);
+  }
+}
+
+/**
+ * copies batch used for creating fixtures onto the blockchain
+ */
+export async function copyBatchForClaim(
+  deployer: string
+): Promise<{ tx: any; postageDepth: number; initialBalance: number }> {
```
Comment (mrekucci) [inline test/util/tools.ts:183]: This reported issue should be fixed.

**Comment by mrekucci** [comment]
Please, resolve the conflicts.

**Comment by mrekucci** [comment]
@nugaon the branch has still conflicts in `src/Redistribution.sol`.

**Review by zelig — APPROVED** [review]

**Pull request merged**